Fix BinTableHDU.copy() memory leak

cmarmo opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address the memory leak in BinTableHDU.copy() identified in #15649.
The code modified in this PR is 11 years old... perhaps numpy behavior changed in the meanwhile so it is possible now to simplify the astropy function? 

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #15649

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk commented on astropy/io/fits/fitsrec.py:
```diff
@@ -581,19 +581,15 @@ def __setitem__(self, key, value):
         return self.names
 
-    def copy(self, order="C"):
+    def copy(self):
```
This changes API - perhaps not an issue?

Also wonder a bit more generally: if `fitsrec` keeps links to some original fits file, then wouldn't the whole file be copied? (This would have been true also with the `deepcopy(self.__dict__)`). It feels like a very large hammer...

cmarmo replied:
To be honest, I kept the deepcopy because of the comment, but it is a bit unclear to me why it was necessary.
The memory duplication seems caused by the `super().copy` plus the deepcopy.

I tried
```
    def copy(self, order="C"):
        new = super().copy(order=order)

        return new
``` 

And all tests passed, so if keeping the API unmodified is a priority and the deepcopy is not necessary, I can commit the version  above.

mhvk replied:
How weird!  So, presumably even just removing `.copy()` would work... (... ... I just tried, all tests pass without the override of `copy()`).

This was all introduced in #1676, but unfortunately without explanation beyond the docstring. I must admit I don't understand why a deepcopy of weird attributes is any better than sharing them. But I think `__getitem__` makes clearer what in fact is the problem: it is all in `_converted`. My sense is that we can use the same technique as in `__getitem__` to get the copy; indeed,
```
# We copy the .view since that will ensure that a separate
# ColDefs is created, and _converted filled accordingly
# (see __getitem__).
new = super(self.__class__, self[...]).copy(order=order)
```
Or, maybe simpler,
```
        # Go through recarray to ensure we do not share items with self
        # that we do not want to share (see __getitem__).
        out = self.view(np.recarray).copy(order=order).view(type(self))
        out._uint = self._uint
        out._coldefs = ColDefs(self._coldefs)
        return out
```

cmarmo replied:
> How weird! So, presumably even just removing `.copy()` would work... (... ... I just tried, all tests pass without the override of `copy()`).

Right... I should have seen it...

I've chosen the simpler solution...

saimn replied:
Some additional info in https://github.com/spacetelescope/PyFITS/commit/9096a66efa99b5111aeaf7b3f9e2bb47a1c467f8:

> This commit contains a fix for a bug that hadn't revealed itself until
making these other changes (but was necessary to fix for all the tests
to pass).  This is specifically the fact that the `FITS_rec.copy()`
method (inherited from `ndarray.copy()`) did not perform a deep copy of
all of a FITS_rec's attributes, including the arrays in the `._convert`
list.  This meant that when copying a `FITS_rec` with the copy method
the new array would have its own copy of the *raw* data, but the
converted arrays would remain shared.  This is especially bad in the new
case where slices of *all* of the columns are stored in `._convert`.
This meant that any update to the new table updated the old table as
well (likewise they shared the same column definitions).  So now
`FITS_rec.copy()` is overridden to ensure that all of the table's
attributes get a deep copy.  This in turn revealed that since
`Column.__new__` expects a minimum number of arguments,
`copy.(deep)copy()` did not work on it without first defining custom
`__(deep)copy__` methods.
> 
> *whew!*

Handling of `_converted` columns is not simple (as far as I remember, but that's `io.fits` so), so maybe the comment helps a bit understanding the original issue. Are we sure that after changing copy's implementation the converted column are not shared with the original table ? (not sure if there is a test for that...)

cmarmo replied:
Thanks @saimn for the additional ref.
I'm going to look for a way to check about converted columns.

cmarmo commented:
... the tests passed locally, I don't understand the reason of the failures  here....

mhvk commented:
The current failures are unrelated - see https://github.com/scientific-python/pytest-doctestplus/pull/242

cmarmo commented:
Sorry for the late answer... life got in the way....

I believe I have added a test checking that the original and copied `_converted` records do not reference the same object...

cmarmo commented:
@saimn , sorry, I'm lost.... as we discussed offline, I'm trying to access the items of the dictionary `_converted` , but in the particular tested case the dictionary appears to be empty... ? ...
There is at least one scaled column in the table: my understanding was that `_converted` is meant to store the scaled values?
I'm surely missing something...

pllim commented:
If this is re-milestone to v7.0 or later, please use the new `perf` change log category if applicable, see:

* https://github.com/astropy/astropy/pull/16694

cmarmo commented:
Hello there... I'm sorry I don't know how to check the items of the `_converted` dictionaries as they both appear empty to me.
There is something I don't understand: can someone help? Thanks a lot for your patience.

saimn commented:
Getting back to this one, the memory leak can be fixed by breaking the references on `_col_weakrefs`:
```
        new = super().copy(order=order)
        new._col_weakrefs = weakref.WeakSet()
        new._coldefs = ColDefs(self._coldefs)
        new._converted = {}
        return new
```
Not sure if resetting `_converted` is needed but I think it's better to be safe. And we would need to test that the BinTableHDU features still work after a copy (e.g. VLA) including for a file read from disk.

cmarmo commented:
@saimn thank you for your patience!
I guess this would go in the 7.x versions, I have updated the changelog entry then.

As you suggested, I have added a test against VLA binary table copies and checked that it was successful on main.
Directly using the [suggested code](https://github.com/astropy/astropy/pull/16143#issuecomment-2442262428) made the test fail with
```
E           OSError: Could not find heap data for the 'QUAL_SPE' variable-length array column.
```

I decided to keep the deepcopy, then, and reset the attributes generating the leak.
Resetting `_converted` is also generating the same `OSError` about heap data, so I only reset `_col_weakrefs`.

saimn commented:
>  Resetting _converted is also generating the same OSError about heap data, so I only reset _col_weakrefs.

Weird but at least we spot that now. I cannot find why so current solution is fine, let's get this fix in.

saimn reviewed (approved):
Thanks @cmarmo !
